File: mobile/lib/features/orders/models/address.dart
class Address {
  const Address({
    required this.fullName,
    required this.phone,
    required this.line1,
    required this.city,
    required this.province,
    this.postalCode = '',
  });

  final String fullName;
  final String phone;
  final String line1;
  final String city;
  final String province;
  final String postalCode;

  factory Address.fromJson(dynamic raw) {
    final json = raw is Map<String, dynamic> ? raw : <String, dynamic>{};

    return Address(
      fullName: (json['fullName'] ?? '').toString(),
      phone: (json['phone'] ?? '').toString(),
      line1: (json['line1'] ?? '').toString(),
      city: (json['city'] ?? '').toString(),
      province: (json['province'] ?? '').toString(),
      postalCode: (json['postalCode'] ?? '').toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'fullName': fullName,
        'phone': phone,
        'line1': line1,
        'city': city,
        'province': province,
        'postalCode': postalCode,
      };
}

File: server/src/models/Address.ts
import mongoose from "mongoose";

const AddressSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true,
      index: true,
      ref: "User",
    },
    label: {
      type: String,
      enum: ["home", "work", "other"],
      default: "home",
    },
    fullName: { type: String, required: true, trim: true },
    phone: { type: String, required: true, trim: true },
    addressLine1: { type: String, required: true, trim: true },
    addressLine2: { type: String, required: false, trim: true },
    city: { type: String, required: true, trim: true },
    state: { type: String, required: true, trim: true },
    postalCode: { type: String, required: true, trim: true },
    country: { type: String, required: true, trim: true, default: "KH" },
    isDefault: { type: Boolean, default: false },
    location: {
      lat: { type: Number, required: false },
      lng: { type: Number, required: false },
    },
  },
  { timestamps: true },
);

AddressSchema.index({ userId: 1, updatedAt: -1 });

export const Address = mongoose.model("Address", AddressSchema);
